extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java

```diff
@@ -83,21 +80,63 @@ public void visitEnd() {
         }
 
         private byte[] hibernateEnhancement(final String className, final byte[] originalBytes) {
-            final byte[] enhanced = enhancer.enhance(className, originalBytes);
+            final byte[] enhanced = enhancerHolder.getEnhancer().enhance(className, originalBytes);
             return enhanced == null ? originalBytes : enhanced;
         }
 
     }
 
     public byte[] enhance(String className, byte[] bytes) {
-        DefaultEnhancementContext enhancementContext = new DefaultEnhancementContext() {
-            @Override
-            public ClassLoader getLoadingClassLoader() {
-                return Thread.currentThread().getContextClassLoader();
+        return enhancerHolder.getEnhancer().enhance(className, bytes);
+    }
+
+    private static class EnhancerHolder {
+
+        private volatile Enhancer actualEnhancer;
+
+        public Enhancer getEnhancer() {
+            //Lazily initialized as it's expensive and might not be necessary: these transformations are cacheable.
+            if (actualEnhancer == null) {
+                synchronized (this) {
+                    if (actualEnhancer == null) {
+                        actualEnhancer = PROVIDER.getEnhancer(QuarkusEnhancementContext.INSTANCE, new ThreadsafeLocator());
+                    }
+                }
             }
+            return actualEnhancer;
+        }
+    }
+
+    private static final class ThreadsafeLocator implements EnhancerClassLocator {
+
+        final ThreadLocal<EnhancerClassLocator> localLocator = ThreadLocal
+                .withInitial(() -> ModelTypePool.buildModelTypePool(QuarkusClassFileLocator.INSTANCE,
+                        CORE_POOL));
```

Comment:
This threadlocal isn't `static` -- is this on purpose?

From what I can see `ThreadsafeLocator` is, in practice, a singleton, so using `static` here would make threadlocal cleanup easier. You know, where we restart in dev mode, drop half the classloaders but keep the same threads, resulting in threadlocals piling up...

Reply:
I would prefer to avoid the `static` actually as that's what makes the Threadlocal depend on the classloader, and therefore causes a whole cathegory of such issues.

To make sure it's not causing a circular reference via threads to classloaders I now manually insepcted it via MAT after performing some cycles of live reload, and it seems fine.

I also more aggreesively clear the threadlocal now - it wasn't strictly necessary but it won't hurt to be cautious, thanks for making me think twice about that :+1: